medusa/restore_cluster.py

```diff
@@ -225,14 +243,29 @@ def _chunk(my_list, nb_chunks):
                 logging.info('Tokenmap is differently distributed. Extra items: {}'.format(extras))
                 topology_matches = False
 
-        if topology_matches:
+        if topology_matches and not self.match_racks:
             # backup and restore nodes are ordered by smallest token and associated one by one
             sorted_backup_nodes = self._tokenmap_to_sorted_nodes(tokenmap)
             sorted_target_nodes = self._tokenmap_to_sorted_nodes(target_tokenmap)
             for i in range(len(sorted_backup_nodes)):
                 restore_host = sorted_target_nodes[i][0]
                 is_seed = True if self.fqdn_resolver.resolve_fqdn(restore_host) in self._get_seeds_fqdn() else False
                 self.host_map[restore_host] = {'source': [sorted_backup_nodes[i][0]], 'seed': is_seed}
+
+        elif topology_matches and self.match_racks:
+            # restore data from the same backup rack into a single rack with matching node count
+            backup_nodes_per_rack = self._tokenmap_to_nodes_per_rack(tokenmap)
+            target_nodes_per_rack = self._tokenmap_to_nodes_per_rack(target_tokenmap)
+            for i in sorted(backup_nodes_per_rack, key=lambda k: len(backup_nodes_per_rack[k]), reverse=True):
```

Comment:
There could be a problem here. NetworkTopologyStrategy will place the next replica on the next node in token order from the next rack. The next rack is picked on alphabetical order, which is not considered here.
That's not a problem as long as nb_racks == rf, but as soon as we have a number of racks different than RF, it will start being a problem.

The way I see it, we should check that after sorting both backup and target racks in alphabetical order, they have the same number of nodes. Associating racks should then be done by that alphabetical order only.
If corresponding racks don't have the same number of nodes, then it means the topologies don't match and we need to fallback to sstableloader.

wdyt?

Reply:
I did not consider this case at all, that's a good point. Also I didn't know about the alphabetical order. I'll update the logic to what you described.